Clone and move different file types and clean up copy assignment files & fix code cov

lwasser opened the pull request:
This will implement a helper function to be consistent with what we move and how throughout #292  

This now also addresses #273 -- i did a small reorg of the template functions. now i need to work on tests as they are failing because of the changes. 

also addresses #172 (indirectly by allowing you to specify things to ignore moving)

Ok this now does a lot of things

1. it removes a bunch of `sys.exit` termination  of kernel to replaces with KeyErrors, etc thaht allow the kernel to keep running
2. It adds a `files_to_grade` wich allows  us to specify what files extentions we want to move from the student clone to submitted for grading
3. It reworks fixtures so  there  is a set o f fixtures that setup a typical classroom directory envt with files an d  assignments and such. this makes tests WAY easier and less redundent. all  new fixtures in conftest so they can  e used across  modules
4.  Added some new  tests specifically for moving files. not done but a solid start

# TODOs

- [x] add docs about `files_to_grade`
- [x] do another check with fresh eyes.

Things are  generally passing now so  i am going to move t his from draft to live and  will revisit next week.  open to any and all  feedback!

lwasser commented on abcclassroom/template.py:
```diff
@@ -197,30 +248,33 @@ def copy_assignment_files(config, template_repo, assignment):
+    #     )
+    #     # TODO - talk to karen about sys.exit vs catching this because
+    #     # in Python sys.exit() locks up the console
+    #     sys.exit(1)
```
@kcranston question for you i spent just a bit of time researching try/ except with sys.exit and sys.exit on it's own but i'm still not clear what best practices are. sys.exit(1) when run in python actually locks up the console totally. i have not figured out how to get my envt back after it runs. 
my questions

1. do you know a way to return to the console command prompt with your envt in tact after sys.exit()?
2. is there any reason why we couldn't just use traditional try/excepts and catch the exception? i just am less familiar with using `sys.exit()` so i figured i would ask.

kcranston replied:
When you say 'console', do you mean in Jupyter? That is unfortunate. Even with try / catch, we often use sys.exit(1) after catching an error. On my local machine (zsh in Terminal), it simply exits the script and leaves my environment intact.

lwasser replied:
@kcranston yes it's odd. i'm working in pycharm on coding. so i have a python console / interpreter that is running the dev envt . yes it really locks everything up. 
it's  fine at the terminal but not fun in vanilla python. i should try python in  bash but i would assume it's the same as python via pycharm. 

have you tried to  run things line by line in an  interperter like pycharm or spyder or something like that?

i wasn't sure if sys.exit() was to halt the command line activities vs try/ catch.

kcranston replied:
Ok, doing some research, it seems that sys.exit (or throwing SystemExit) may kill the kernel, so that's not good.

lwasser replied:
ok thank you for the sanity check. it's been driving me a bit batty having to rebuild my envt each time when working on things! So... is a traditional try / except approach better? that is what i'm more familiar with.  when i tried it in the template module, it printed all of the traceback and then my message

lwasser commented on abcclassroom/template.py:
```diff
@@ -265,11 +324,17 @@ def create_extra_files(config, template_repo, assignment):
         for f in extra_path.iterdir():
-            print(" {}".format(f.relative_to(course_dir)))
-            shutil.copy(f, template_repo)
+            if not ntpath.basename(f) in files_to_ignore:
```
here `ntpath` supposedly ensures that this will work on windows as well... not tested but i read that was better than using os.path to check `basename`. open to any feedback

kcranston replied:
If we ensure that f is a pathlib object, and not a string, you can use `name`  https://docs.python.org/3/library/pathlib.html#pathlib.PurePath.name and be safe on all platforms

lwasser replied:
oh! ok i'll try `name`. we should be working with pathlib objects here.. i'm pretty sure i made sure that was the case. i went down a SO rabbit hole on paths and odd quirks on linux and windows... maybe i went too far. :)

lwasser replied:
this one is really tricky. Pathlib.glob() method returns a generator object which in ptyhon you iterate through it once and then it's empty if you try to use it again. you can call list() on it and it will become a list b ut then you can't use the methods pathlib provides. but something odd i s happening when you iterate through the generator  because it seems like it's producing strings each time rather than a path object which means i can't use `.name` will revisit this as well in the future. it seems to be working now as i wrote it but worth a note here.

lwasser replied:
is there any reason to not just use `shutil.copytree` here? with the ignore option? that seems like it will handle everything that we want it to handle even directories and cut back on code

lwasser commented:
i'm getting a weird windows regex failure 

https://github.com/earthlab/abc-classroom/pull/295/checks?check_run_id=1216620744#step:5:106 

the expected message is right - does anyone know anything about windows and regex by chance? @nkorinek @kcranston ??

kcranston commented on abcclassroom/template.py:
```diff
@@ -141,14 +191,12 @@ def create_template_dir(config, assignment, mode="fail"):
-                before re-running""".format(
-                    template_path.relative_to(course_dir)
-                )
+            raise Exception(
```
I would raise a specific exception rather than just Exception: https://docs.python.org/3/library/exceptions.html#FileExistsError

kcranston replied:
Then, you can just check that you get the right kind of exception raised when the dir already exists, rather than a regex against the error message. (and I have no idea why that regex is failing on windows!)

lwasser replied:
totally...i can definitely test to see what the specific exception is and catch it. but `match=` will test that the message you expect is actually there - in other words that you catch the exception that you think you are catching.

it's super odd that windows is not happy with this check however???!!!

thank you for this input karen!!

lwasser replied:
ok i figured it out. it was a path issue so i just removed it testing the specific path return. i have a few thoughts on what it was but am out of energy to problem solve anymore. i will work on the exception type later just want to first get this all running.

lwasser commented on abcclassroom/tests/test_assignment_template.py:
```diff
@@ -85,11 +85,9 @@ def test_create_template_dir_fail_when_exists(default_config, tmp_path):
+
     # If run again, it should fail given the dir already exists
-    with pytest.raises(Exception, match=("Directory " + expected_path)):
+    with pytest.raises(Exception, match=("Oops! The directory specified")):
```
Here i was matching the text + the path and it  caused an issue only on windows. no need to do that given we just want to make sure our error and our associated message is actually being presented to the user!

kcranston replied:
We should also raise a `FileExistsError` in `template.py`, and then write this test as `pytest.raises(FileExistsError, match=("Oops! The directory specified"))`. Thinking ahead to using this as a library, and throwing an exception specific to the type of error that we encounter for future clients.

lwasser replied:
totally. i hadn't dug into that yet because i was confused by the fail mode and how / when it's triggered. so i was going to look at that next. i was also confused by why delete mode keeps the .git directory! if you could explain that it would be helpful! once i understand the modes better it will be easier to add checks!

kcranston replied:
The `fail` mode is the default for the `abc-new-template` script. That mode is not available for `abc-update-template`. As for the `.git` dir, this is tricky. There is some explanation at https://github.com/earthlab/abc-classroom/blob/master/abcclassroom/template.py#L169. Can explain more, but its complicated and this is not the best venue!

lwasser commented:
wow windows tests via github actions are WAY faster than appveyor and it was way easier to setup!!  yay!!

lwasser commented:
heck yea passing tests!!

lwasser commented on abcclassroom/template.py:
```diff
@@ -27,33 +28,82 @@ def new_update_template(args):
-    github.init_and_commit(template_repo_path, args.custom_message)
+    # If the assignment exists, then create dirs and copy stuff
+    template_repo_path = create_template_dir(config, assignment_name, mode)
+    # This is copying DS_Store still
```
```suggestion
```

nkorinek commented:
@lwasser I modified the`copy_assignment_files` function to use the `utils.copytree` function to copy over whatever type of files the user wants to copy over. I wasn't sure how to add it as a config option, so that may be the next step. For now I just added it as a parameter in the function.

lwasser commented on abcclassroom/clone.py:
```diff
@@ -99,17 +99,58 @@ def clone_student_repos(args):
 
 
-def copy_assignment_files(config, student, assignment_name):
+def include_patterns(patterns):
```
@nkorinek i am working on the clone module in #318 so just be careful here but this definitely should go in utils.

lwasser commented:
ok @nkorinek 
how about you move that function to utils and then look at how i implemented files to ignore here:
https://github.com/earthlab/abc-classroom/commit/d0f6753cdd8d8627bce7d125311bb4b5df230b10#diff-83f040abdd4838190f6e6f4a833ff1c44e5e35ccf8a004b114966d7039b7bd27R72 

NOTE - my implementation is a hack. but essentially i thought  that the helper could just grab the files from the config (this works) and subset out what it should ignore based upon the config. 
THen you don't have to add a parameter to the helper.

The helper could do the same with files_to_move but we may want to chat about that more to avoid making the config too complete. but why don't you start with implementing the files_to_ignore?

lwasser commented on abcclassroom/clone.py:
```diff
@@ -168,36 +168,31 @@ def copy_assignment_files(config, student, assignment_name):
         config file returned as a dictionary from get_config()
-    student:
+    student: string
+        Name of the student who's files are being copied
```
```suggestion
        Name of the student whose files are being copied
```

lwasser commented on abcclassroom/clone.py:
```diff
@@ -168,36 +168,31 @@ def copy_assignment_files(config, student, assignment_name):
+    print("Copying files from {} to {}".format(Path(source_dir), destination))
+
+    # Using the copytree function from util to make copying easier
+    shutil.copytree(
```
ok i just played with this. it does work however because it is copying everything i had to add a bunch of things to files_to_ignore. some of the things in files_to_ignore i'd want to move in the template  move so maybe this one function handles moving files differently than the 3 other places??

lwasser commented:
ok after playing with this a lot i do think it's important to be able to specify what types of files we want to move to the submitted dir after cloning. I also think in the long run IF we want to move images - we may want to consider removing any files that exist in the extra_files directory. but for now this will serve our immediate course use case.

lwasser commented on abcclassroom/tests/test_assignment_template.py:
```diff
@@ -3,31 +3,77 @@
+        "release",
+        assignment_name,
+    )
+    release_path.mkdir(parents=True)
```
i  thought i rewrote this -- TO DO  go  back and double check this part

lwasser commented on abcclassroom/tests/test_clone.py:
```diff
@@ -103,23 +100,74 @@ def test_roster_wrong_format(sample_course_structure):
+        abcclone.copy_assignment_files(config, a_student, assignment_name)
+
+
+# def test_copy_assignment_files_optional_files(
```
TODO -- re work this with the new fixtures!

lwasser commented on docs/collect-grade-assignments/clone.rst:
```diff
@@ -57,6 +57,14 @@ files into the nbgrader ``submitted`` directory using the structure:
 The path to ``course_materials`` is defined in ``config.yml`` file. ``abc-clone``
 will create subdirectories within ``course_materials`` for each student as needed.
 
+Cone and Do Not Move to Submitted
```
```suggestion
Clone and Do Not Move to Submitted
```

lwasser commented:
OK! waiting to look at code cov and tests but this PR now does a lot of things

1. i've tested using  abc-clone and abc-new-template as a package and made a few tweaks to streamline. it works great as a .py script. i'll add some documentation around this as i refine the workflow but it's all working
2. i've revamped conftest to include a set of fixtures that pretty much setup the abc and nbgrader  course structure.  this will make writing tests WAY WAY easier. it's really nice  now. 
3. i've added a  skip moving to submitted flag that can be used if you just want to update clones but not mess  with grading files. this seems reasonable because we can  also revert a repo back to a previous state so i'm not worried about messing with the cloned repos with a pull. i am   worried about messing with files we grade (altho with git that could be easy to resolve we sometimes make fixes to notebooks to make them run).
4. i've made some small changes to the docs  and  wrote a few new tests. 

There is still a good bit of cleanup to do  here and a good bit of tests  to write  but i think this PR is read enough too  merge. 

i'll work on the authentication next so  we can merge and bump a new version!

lwasser commented:
odd - i'm noticing t hat code cove  has  not been tracking recent commits. will look into this now.

lwasser commented:
Ok  i ALSO updated code cov which wasn't updating given we moved from travis. I had to do a few things

1. I did add a token  to  secrets even tho it says we dont need it for public repos 🤷 
2. I also made  sure i had the most recent version of this action which  helped. 
3. I removed the matrix build and am only building in linux.

NOTE:  it could  be possible to bundle this into another build but  i'm not sure  how to do that and not have it run on 3 different  python versions and push three times so for now,  we will just use t ihs.
